### PR TITLE
Improve Code Quality

### PR DESCRIPTION
**Pull request by bhgomes**
This PR is based on the `edition` branch https://github.com/rust-zkp/ark-plonk/pull/37

## Changes
- fixed `clippy` suggestions or marked where `clippy` is wrong
- fixed broken doc links
- removed `get_domain_attrs` in favor of `EvaluationDomainExt`
- removed unnecessary import renaming
- general best-practices improvements

### CODE REVIEW

File: src/proof_system/proof.rs
Comment (davidnevadoc) [inline src/proof_system/proof.rs]: Yes, that's better

**Review by davidnevadoc — COMMENTED** [review]
Now that we are using `domain.group_gen()` it would be good to use it in place of `domain.element(1)` ( [proof.rs/L247](https://github.com/rust-zkp/ark-plonk/blob/a8110966b777a2eaf40556587814ec9189598ab0/src/proof_system/proof.rs#L247)) These calls return the same element . However I think using the first option is clearer and more consistent.

File: .github/workflows/ci.yml
Comment (CPerezz) [inline .github/workflows/ci.yml:83]: Not sure why we need `workspace` when this is a single crate. Could you elaborate on that?

File: src/circuit.rs
Comment (CPerezz) [inline src/circuit.rs:307]: We can maybe move this to the `lib.rs` if we thing will pop up more.

File: src/constraint_system/composer.rs
Comment (CPerezz) [inline src/constraint_system/composer.rs:30]: Why do we want this feature? And also, in case is needed? Shouldn't we enable it at `Cargo.toml` level with a feature-flag like it's done for `parallel` and `asm`?

File: src/constraint_system/ecc/scalar_mul/fixed_base.rs
Comment (CPerezz) [inline src/constraint_system/ecc/scalar_mul/fixed_base.rs]: Edit the coment and fix it directly instead of leaving the FIXME please. :)

File: src/constraint_system/ecc/scalar_mul/fixed_base.rs
Comment (CPerezz) [inline src/constraint_system/ecc/scalar_mul/fixed_base.rs]: We should move this lint to `lib.rs` and allow it for the entire lib. As sometimes we want to use capital letter variables.

File: src/constraint_system/ecc/scalar_mul/fixed_base.rs
Comment (CPerezz) [inline src/constraint_system/ecc/scalar_mul/fixed_base.rs]: ditto.

File: src/error.rs
Comment (CPerezz) [inline src/error.rs]: Shouldn't we then leave an issue waiting for the PR instead of adding code that we don't yet need directly?

File: src/permutation/constants.rs
Comment (CPerezz) [inline src/permutation/constants.rs:10]: Remove this from here and just add it in `lib.rs` to avoid repetition.

File: src/permutation/permutation.rs
Comment (CPerezz) [inline src/permutation/permutation.rs:974]: I would prefer to avoid touching anything related to domain internal field obtentions until https://github.com/arkworks-rs/algebra/pull/333 is released.

This will make it much more easy and organic. But for now I would prefer to get all of the elements on the same way. Instead of some on one way, and some on the other..

File: src/permutation/permutation.rs
Comment (CPerezz) [inline src/permutation/permutation.rs:1113]: ditto.

File: src/permutation/permutation.rs
Comment (CPerezz) [inline src/permutation/permutation.rs:1270]: ditto.

File: src/proof_system/linearisation_poly.rs
Comment (CPerezz) [inline src/proof_system/linearisation_poly.rs]: ditto.

File: src/proof_system/preprocess.rs
Comment (CPerezz) [inline src/proof_system/preprocess.rs]: ditto.

File: src/proof_system/proof.rs
Comment (CPerezz) [inline src/proof_system/proof.rs:485]: ditto for `domain.size_inv()`

File: src/proof_system/proof.rs
Comment (CPerezz) [reply src/proof_system/proof.rs]: I don't agree. Until the PR we made to arkworks does not land, I'd not mix the two ways of getting data from the struct TBH. I know the `util::` is ugly, but if we have the two ways at the same time makes it really confusing.

File: src/proof_system/prover.rs
Comment (CPerezz) [inline src/proof_system/prover.rs:95]: ditto.

File: src/util.rs
Comment (CPerezz) [inline src/util.rs:119]: That is basically replaced by https://github.com/arkworks-rs/algebra/pull/333/files as soon as is released. So I won't merge it here. Specially considering the repo should ideally contain only PLONK-related logic and avoid adding any external stuff as much as possible (the solution is really good obviously) But I prefer to wait for the new version rather than have a lot of new code that we will need to remove soon anyway.

**Review by CPerezz — CHANGES_REQUESTED** [review]
Thanks for the PR!

Two main things.
- I would delete everything related to `EvaluationDomainExt`. You're trying to solve in this crate a design issue of a different crate. Which has already been addressed and is waiting for a release. See: https://github.com/arkworks-rs/algebra/pull/333 (It's basically the same idea you coded :smile: )
- I'm not a fan of the removal of the `PlonkVerifierKey` alias. And the reason is that in `ark-poly-commit` there are quite some structs/traits called on the same way. And in advance to that, we changed the name of our own structure to make the distinction of what's PLONK-related and what's polynomial-commitment related. So personally, I would prefer to un-do it.

Other than that a few nits and suggestions. But thanks a lot for your contribution @bhgomes  we really appreciate it! Sorry for the comments and change requests...

File: src/proof_system/proof.rs
Comment (davidnevadoc) [reply src/proof_system/proof.rs]: In any case, we should use `group_gen_inv` and remove the `inverse()` call

File: .github/workflows/ci.yml
Comment (bhgomes) [reply .github/workflows/ci.yml:83]: `--all` is the deprecated form of `--workspace` so I was just bringing it up to date.

Also, note that I was able to push changes to the CI which is not something non-contributors should be allowed to do. You should probably change the permissions settings so that only core contributors can change actions files.

File: src/circuit.rs
Comment (bhgomes) [reply src/circuit.rs:307]: NB: I meant `harsh` there instead of `hash`.

It's generally not a good idea to dismiss `clippy` lints globally unless they are really broken or too strict like `pedantic`. Lints related to code complexity are important for constraining the design space and sometimes `clippy` is "wrong" in that it treats all complexity the same using a length heuristic, when that's not always what makes something complex or not. See for example other parts of the library where `FIXME` was the note instead for complexity-related lints. It's important to distinguish these two cases when we know the readability and design are not ideal, and when `clippy` is just not subtle enough.

File: src/constraint_system/composer.rs
Comment (bhgomes) [reply src/constraint_system/composer.rs:30]: This is there because those are only imported when `trace` is enabled. I think the `Cargo.toml` is a bit strange here with the `trace` flag and the `trace-print` flag. Usually, you want features to group dependencies or large sections of a library but in this case it seems to be that `trace` is used for only one function and doesn't relate to any dependencies. Maybe this function `check_circuit_satisfied` makes sense to return a boolean, but certainly, it is a little confusing. I hadn't seen this in my code review.

File: src/constraint_system/ecc/scalar_mul/fixed_base.rs
Comment (bhgomes) [reply src/constraint_system/ecc/scalar_mul/fixed_base.rs]: I wrote this because I didn't know what the new comment should be.

File: src/constraint_system/ecc/scalar_mul/fixed_base.rs
Comment (bhgomes) [reply src/constraint_system/ecc/scalar_mul/fixed_base.rs]: Honestly, I'm of the opinion you should not use non-Rust conventions for naming and it should only be used in cases where you are trying to match some specification in which case it's more important to get the spec and impl as close to each other as possible. Like code complexity lints, it should be disabled in a more precise way than just everywhere. I just did the smallest change here to remove the issue.

File: src/error.rs
Comment (bhgomes) [reply src/error.rs]: Sorry, that was an oversight. I meant to delete that before pushing.

File: src/permutation/constants.rs
Comment (bhgomes) [reply src/permutation/constants.rs:10]: Disabling lints in a precise way is not repetition, as each one should be used to indicate some important convention in that region of code. Naming lints can also be used as compile-time sanity checking.

**Comment by bhgomes** [comment]
> I would delete everything related to `EvaluationDomainExt`. You're trying to solve in this crate a design issue of a different crate. Which has already been addressed and is waiting for a release. See: [arkworks-rs/algebra#333](https://github.com/arkworks-rs/algebra/pull/333) (It's basically the same idea you coded 😄 )

There are several reasons to make this change now:
1. Arkworks is slow to release new code, especially for `crates.io` releases. This library is moving much faster and should not wait for them.
2. Even if it is *necessitated* by a design issue in their crate, the old solution is a design issue in this crate. It doesn't matter that they did it wrong, we shouldn't do it wrong here too. The old solution is not good for many reasons as I describe above.

Also, a few times in the comments above you refer to:
> I know the `util::` is ugly, but if we have the two ways at the same time makes it really confusing.

Not sure what you mean by "two ways". This PR removes that method entirely and actually reduces the number of ways that these kinds of data are being accessed.

> I'm not a fan of the removal of the `PlonkVerifierKey` alias. And the reason is that in `ark-poly-commit` there are quite some structs/traits called on the same way. And in advance to that, we changed the name of our own structure to make the distinction of what's PLONK-related and what's polynomial-commitment related. So personally, I would prefer to un-do it.

The alias was not used as an alias in those files where I changed it back. Aliases are used to distinguish between two objects that have the same name but refer to different things. When you need to use an alias consistently, this may be an indicator of bad naming and/or design. "If the name should be changed, why not change the name of the type itself instead of every time you import it." My preferred solution would be to refer to them as `VerifierKey` and `kzg10::VerifierKey` instead of the convention used currently in those two files. Also, this is the `ark-plonk` crate and anytime someone sees a name that is not qualified with a module namespace, the first assumption they will make is that it belongs to the entire crate, which is a PLONK crate, so `VerifierKey` should be the PLONK `VerifierKey`, not some part of the implementation. Also, I think in general, the way that `ark-poly-commit` does it is the correct way.

File: src/proof_system/prover.rs
Comment (bhgomes) [reply src/proof_system/prover.rs:95]: See above.

File: src/permutation/permutation.rs
Comment (bhgomes) [reply src/permutation/permutation.rs:974]: Not sure what you mean by "some one way and some on the other". This PR removed the entire `util::get_domain_attrs` method.

File: src/util.rs
Comment (bhgomes) [reply src/util.rs:119]: Here are some reasons why this should be kept:
1. This repo is already containing non-PLONK logic. This is the nature of software and we cannot expect our code to be a spec implementation only.
2. Using `&'static str` to query members of an object cannot be checked by the compiler and so when we do something wrong like mistype the name we cannot know if it is wrong unless we have a test that passes through that line and even then you might not have a good way of finding that bug. Getting full code coverage is harder than just using extension `trait`s.
3. Using extension traits to extend non-local libraries is standard practice in Rust so other developers will recognize this design language and understand its purpose more readily.
4. Using a trait allows for better discoverability and documentation.
5. Using a trait allows for abstraction over implementations.
6. If/when `arkworks` decides to release the new version, the change here is just to delete the extension trait, everything else stays the same.

File: src/proof_system/preprocess.rs
Comment (bhgomes) [reply src/proof_system/preprocess.rs]: Note how this lint indicates a change to be made rather than an oversight on the linter. Those two kinds of situations are important to distinguish between. See my comments above about linting in general.

File: src/proof_system/linearisation_poly.rs
Comment (bhgomes) [reply src/proof_system/linearisation_poly.rs]: Note how this lint indicates a change to be made rather than an oversight on the linter. Those two kinds of situations are important to distinguish between. See my comments above about linting in general.

**Comment by bhgomes** [comment]
> Thanks for the PR!
>
> ...
>
> Other than that a few nits and suggestions. But thanks a lot for your contribution @bhgomes we really appreciate it! Sorry for the comments and change requests...

Of course, happy to contribute! :)

**Comment by bhgomes** [comment]
> The alias was not used as an alias in those files where I changed it back. Aliases are used to distinguish between two objects that have the same name but refer to different things. When you need to use an alias consistently, this may be an indicator of bad naming and/or design. "If the name should be changed, why not change the name of the type itself instead of every time you import it." My preferred solution would be to refer to them as `VerifierKey` and `kzg10::VerifierKey` instead of the convention used currently in those two files. Also, this is the `ark-plonk` crate and anytime someone sees a name that is not qualified with a module namespace, the first assumption they will make is that it belongs to the entire crate, which is a PLONK crate, so `VerifierKey` should be the PLONK `VerifierKey`, not some part of the implementation.

Also, on the matter of `VerifierKey`s: there seems to be an issue with the design of this abstraction in general in this crate, so some more careful thinking should go into a final solution for this case.

File: src/constraint_system/composer.rs
Comment (CPerezz) [reply src/constraint_system/composer.rs:30]: Ohh that was my bad. I missread and thought BigNumber was added as dep. And therefore wanted to constrain it's usage via features in `Cargo.toml` directly.

File: src/constraint_system/ecc/scalar_mul/fixed_base.rs
Comment (CPerezz) [reply src/constraint_system/ecc/scalar_mul/fixed_base.rs]: Something like:
```
    // # Note
    // This function is optimized for fixed base ops **ONLY** and therefore,
    // the **ONLY** input that should be passed to the function as a point is the
    // generator or basepoint of the curve over which we are operating.
```
should do. :smile:

File: src/constraint_system/ecc/scalar_mul/fixed_base.rs
Comment (CPerezz) [reply src/constraint_system/ecc/scalar_mul/fixed_base.rs]: So while I understand your POV, and i respect it, with EC libs, sometimes it's much much easier to read code when you can identify variables and differenciate scalars and points just by seeing whether they're capital or non-capital letters. 

Let's keep the lint for now. But I'm slightly biased on this one to move the lint to be general for the entire library..

File: src/error.rs
Comment (CPerezz) [reply src/error.rs]: Can you then delete it please? Or you left it due to any other reason? :)

File: src/proof_system/linearisation_poly.rs
Comment (CPerezz) [reply src/proof_system/linearisation_poly.rs]: I understand your point. But I do also consider the priorities of the lib.

I agree, if we can afford reducing the complexity of the types and fix this would be nice. But that should never be in charge of increasing the complexity of the implementation complexity itself. As the algorithm is already complex and difficult sometimes to match to the paper. So if to satisfy clippy and some Rust standards we need to deviate from the implementation style a lot, I would not feel it's worth TBH.

That was my concern when I pointed the clippy lint complaints.

File: src/proof_system/preprocess.rs
Comment (CPerezz) [reply src/proof_system/preprocess.rs]: Ditto response above.

File: src/util.rs
Comment (CPerezz) [reply src/util.rs:119]: After discussing with @davidnevadoc he agreed on leaving the struct/trait that you purposed. 

So fine from my side then.

File: src/error.rs
Comment (bhgomes) [reply src/error.rs]: I was just waiting for your comments to do a commit.

**Comment by CPerezz** [comment]
Hi @bhgomes .

First, after discussing with @davidnevadoc agreed on leaving the Domain trait solution until the PR is shipped in a new `arkworks ` version.

Second, on the `clippy` lints. I pointed them out as it kinda leaves the door open to a refactor of quite some types. 
And while I'm not against it. As said in the comments, we all value more code clarity and similarity to the paper rather than rust standards compliance.

For "said in the comments", I refer to:
>I understand your point. But I do also consider the priorities of the lib.
>
>I agree, if we can afford reducing the complexity of the types and fix this would be nice. But that should never be in charge of increasing the complexity of the implementation itself. As the algorithm is already complex and difficult sometimes to match to the paper. So if to satisfy clippy and some Rust standards we need to deviate from the implementation style a lot, I would not feel it's worth TBH.

That was my concern when I pointed the clippy lint complaints.


For the rest I think that's all. If you can remove the `Result` thingm I think we can merge this. Then I'll edit the CI config change permissions :)

**Review by CPerezz — APPROVED** [review]
Approved modulo de removal of the `Result` type inclusion.

**Comment by bhgomes** [comment]
@CPerezz Thanks for your feedback! I'll make the relevant changes later today.

As for the `clippy` lints, I'll try to find a reasonable compromise since it is important to match the spec and to do so without obfuscation. I'd like to do a more thorough review later anyways to check for deviations from the spec, at least superficially.

**Review by LukePearson1 — APPROVED** [review]
Top code quality @bhgomes, I agree with the trait abstractions. LGTM overall.

On a side note: good to see more people getting involved in open source work.

**Review by davidnevadoc — APPROVED** [review]
LGTM 👍

**Pull request merged**
